### ISSUE TITLE
Bucket name length limitations in GCS

### ISSUE THREAD

**Issue by lukeweber** [crossplane/crossplane#458]
When generating bucket names in crossplane in GCS we have a "name-uuidv4" combo. In this case the limit on the name would be 26 chars because a uuid is 36 + "-" separator. If we're generating names, we could consider adding a ".uuid" instead of a - to allow 63 char names.

From: https://cloud.google.com/storage/docs/naming
"Bucket names must contain 3 to 63 characters. Names containing dots can contain up to 222 characters, but each dot-separated component can be no longer than 63 characters."

**Comment by hasheddan** [comment]
@lukeweber @ichekrygin I would be happy to take this one if no one else has yet!

**Comment by ichekrygin** [comment]
> When generating bucket names in crossplane in GCS we have a "name-uuidv4" combo.

That is **only** if the user specifies bucket name format as such:
```yaml
apiVersion: storage.gcp.crossplane.io/v1alpha1
kind: Bucket
metadata:
  name: demo-bucket
  namespace: crossplane-system
spec:
  #...
  nameFormat: bucket-name-%s
```

If the user does not specify `nameFormat`, bucket name will default to UUID. 

If the user specifies a particular name say: `nameFormat: my-special-bucket`  or `nameFormat: my.special.bucket`, then that will be the name of the resulting bucket. 

GCS buckets names are globally unique; hence, there must be some sense of coordination. If the user doesn't care what the bucket name is, then either `prefix-UUID` or `UUID` works just fine. If the user cares about the resulting bucket name, then they can choose that appropriately. In this sense, naming functionality/requirements of the GCS buckets is not that much different than in `gsutil` tool.

With the existing functionality, user can specify `nameFormat: [bucketName].%s` , which will allow `bucketName` value to be as long as specified in the documentation, i.e., 233 chars. With that said, I don' think this is a good and valid approach because by using `dot` separator has additional ramifications:

> Names containing dots require [verification](https://cloud.google.com/storage/docs/domain-name-verification)

Taking into consideration the random value nature of `UUID` values, I don't think it is practical (or feasible) to expect UUID value to pass domain name verification.

To sum up, with existing functionality of `nameFormat: value` user can create buckets with names containing `dot` separator, with or without `UUID` suffix.

**Comment by lukeweber** [comment]
@ichekrygin - Do you think it might be valuable/useful to validate the nameFormat for length, and error before hitting the api when we know it's going to fail. We have the same issue with s3 buckets as well.

**Comment by ichekrygin** [comment]
@lukeweber yes, always! What would be helpful is an example or `regex` for schema validation to support use case:
`value.dot.%s.delimited`: total length < 223 chars, each dot delimited segment < 64 chars
- in this case, `%s` needs to be counted as a full length of k8s UUID valid.

I don't think the `domain validation` aspect should be in-scope for the validation.

**Comment by lukeweber** [comment]
Api machinery might have some validation that's helpful in this pursuit, since they do validate service names and other types of objects  in an similar way in kubernetes. https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go

**Comment by lukeweber** [comment]
@hasheddan - You're welcome to this one.

**Comment by ichekrygin** [comment]
> Api machinery might have some validation that's helpful in this pursuit, since they do validate service names and other types of objects in an similar way in kubernetes.

@lukeweber are you suggesting the "validation" that extends beyond schema validation?

**Comment by ichekrygin** [comment]
Also, just for the reference, @lukeweber could you please indicate what was the "Failure" condition message when the bucket name exceeded given length?

**Comment by ichekrygin** [comment]
@hasheddan I don't oppose you starting on this issue, however, it would be very helpful to gain some clarity on the approach. Since there are "multiple" ways to solve this (just like almost anything)  - some ways are more preferable than others. For example, I would hope that we avoid adding said "validation" to the controller itself.

**Comment by jbw976** [comment]
@ichekrygin do you think this validation can be done cleanly and simply at the CRD schema level? i.e. https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#validation

That would not add any code to our controllers, it would simply be declarative and enforced by k8s api-server.

**Comment by negz** [comment]
As of https://github.com/crossplane/provider-gcp/pull/208 GCS is switched over to our `crossplane.io/external-name` annotation, and no longer uses `NameFormat`. Under this pattern the bucket names should typically be a lot shorter. If you omit the `crossplane.io/external-name` annotation your bucket will be named `claimnamespace-claim-name-5char`, for example `default-coolbucket-dk4jd`. If you specify the annotation, your bucket will get the specified name. For example a bucket claim annotated `crossplane.io/external-name: mycoolbucket` would be named `mycoolbucket`.

> Do you think it might be valuable/useful to validate the nameFormat for length, and error before hitting the api when we know it's going to fail.

Our approach in Crossplane for some time has been to let folks submit invalid requests to the underlying API and be sure to clearly return any error from the underlying API. It's simply a losing game to try to keep our code (or even CRD schemas) updated with all the possible valid and invalid values that each API supports.
